### DISCUSSION TITLE
Is a Custom Folder Structure Possible?

### DISCUSSION THREAD

**Discussion by jameshibbard** [beetbox/beets#5551] [General]
I would like to create with  a custom folder structure for my media. For example:

```
.
├── Audio Books
├── Electronic
│   ├── Techno
│   └── Trance
├──  Genres
│   ├── Rap
│   ├── Jazz
│   └── Metal
└── Singles
```

How would I go about approaching this?

For example, having the following in my config.yaml:

```
paths:
  default: Genres/$genre/$albumartist/$year - $album/$track - $title
```

Means that Wu-Tang Clan is imported to the correct folder (`Genres/Rap`). But running `beet import Powerslave` (an Iron Maiden album) results in a "Heavy Metal" folder being created in `Genres`, which is not what I want. 

Instead I would like each sub genre of metal to be tagged as such (e.g. tagged as "Heavy Metal"), but to be put into the `Genres/Metal` folder

Also, how would I go about importing into `Singles` or `Audio Books`? 

Or what about when an album cannot be found by the MusicBrainz lookup? In this case selecting "Use as-is" results in it being put into `Genres`.

Sorry for the unfocussed question, but I would be grateful if someone could point me to where to look. Ideally, it would be great if I could define some `path` rules in my config file and pass a command line flag (or whatever) to indicate which rule to apply on import.

Grateful for any advice.

**Comment by jameshibbard** [comment]
After spending the last couple of days reading the docs and combing the interet, I am in a position to answer my own question. I am posting this here in case it helps anyone else.

> Is a Custom Folder Structure Possible?

Yes, it is. However, the first thing to realize is that filenames and paths are generated by rules from fields. You don’t modify them manually.

If you're unsure about the desired folder structuure of your music library, it might be good to read this: https://www.reddit.com/r/musichoarder/comments/1b9boeo/folder_structure/

> How would I go about approaching this?

A high-level answer would be to look at [flexible attributes](https://beets.readthedocs.io/en/stable/guides/advanced.html#store-any-data-you-like) and the [inline plugin](https://beets.readthedocs.io/en/latest/plugins/inline.html). The [move command](https://beets.readthedocs.io/en/stable/reference/cli.html#move) could also be useful.

> I would like each sub genre of metal to be tagged as such (e.g. tagged as "Heavy Metal"), but to be put into the Genres/Metal folder.

You can set a flexible attribute when importing an album, then create a path rule to stick it in the right place. E.g.:

```
import --set="mygenre=metal" Powerslave
```

And:

```
paths:
  mygenre:metal: Genres/Metal/$albumartist/$year - $album/$track - $title
```

> Also, how would I go about importing into `Singles` or Audio Books?

For singles you can run in singleton mode using `beet import -s <whatever>`. This tags individual tracks instead of whole albums at a time. You can define a separate rule for singletons, but the defaults look like this:

```
paths:
  default: $albumartist/$album%aunique{}/$track $title
  singleton: Non-Album/$artist/$title
  comp: Compilations/$album%aunique{}/$track $title
```

Ref.: https://beets.readthedocs.io/en/stable/reference/config.html#path-format-configuration

As for `Audiobooks` import them using a flexible attribute (as above) and define a path rule.

> Or what about when an album cannot be found by the MusicBrainz lookup? In this case selecting "Use as-is" results in it being put into `Genres`.

The cleanest solution (as far as I can tell) is to use a flexible attribute. However, you can also move folders within your library using: 

```
beet move -d /home/jim/Music/Metal/ "Powerslave"
```

I also mentioned the inline plugin. This lets you use Python to customize your path formats. I found this useful, as if I have multiple albums by one artists, I want the folder structure to be like this:

```
├── Metal
    └── Iron Maiden
        ├── 1981 - Killers
        └── 1984 - Powerslave
```

Whereas, if I only have one album by a particular artist, I don't want to create a dedicated artist folder, rather I want:

```
├── Metal
    ├── Black Sabbath - Paranoid (1970)
    └── Iron Maiden
        ├── 1981 - Killers
        └── 1984 - Powerslave
```

This can be achieved using the inline plugin like follows:

```
plugins: ... inline

album_fields:
  one: |
	album_count = globals().get("album_count")
	if album_count == "1":
	  return f"{albumartist} - {album} ({year})"
	else:
	  return f"{albumartist}"
  two: |
	album_count = globals().get("album_count")
	if album_count == "1":
	  return f""
	else:
	  return f"{year} - {album}"

paths:
  mygenre:metal: Metal/$one/$two/$track - $title
```

And albums should be imported like so:

```
beet import --set="mygenre=metal" "Killers"
beet import --set="mygenre=metal" "Powerslave"
beet import --set="mygenre=metal" --set="album_count=1" "Paranoid"
```

If, at any point, I obtain a second Black Sabbath album, then removing the `album_count` attribute on `Paranoid` will cause it to be saved in a dedicated artist folder. beets will rewrite the path automatically.

```
beet modify album_count! -a "Paranoid"
```

Ref.: https://beets.readthedocs.io/en/stable/reference/cli.html#modify

---

I hope this helps someone. I would also be happy to hear if there is a better way to do all of this.

**Reply by wisp3rwind** [reply]
> I hope this helps someone. I would also be happy to hear if there is a better way to do all of this.

Flexible attributes and the inline plugin for path formats are generally the right tool to achieve custom folder structures (beets mostly operates under the assumption of being in charge of managing the folder structures: You can tell it to not touch file paths at all, but that's not really the primary mode of operation.)

Since your "custom" genres (i.e. super-genres like "Metal") are not available directly in the metadata that beets fetches, you'll have to cook up a custom solution like you did.
An alternative (but not necessarily better) way to achieve the same outcome might be to create an inline field that uses some heuristics to come up with these genres:

```yaml
inline:
  album_fields:
    mygenre:
      genre = genre.lower()
      if "metal" in genre:
        return "Metal"
      elif "rap" in genre::
        return "Rap"
      else:
        return genre
```

Then, you'd not need to set the `mygenre` attribute on each import (probably at the cost of sometimes needing to adjust the heuristics in the inline field).

**Reply by jameshibbard** [reply]
Thank you for the reply. It's nice to know I'm not doing it all wrong.

> You can tell it to not touch file paths at all

How would you do this? Do you mean using `beet import -C ...`

> An alternative (but not necessarily better) way to achieve the same outcome might be to create an inline field that uses some heuristics to come up with these genres:

Nice. That will save me some manual work. I haven't fully decided on the folder structure, as going down this rabbit hole has made me think long and hard about how I want things to be organized. But this is definitely something I will have a look at.

**Reply by wisp3rwind** [reply]
> > You can tell it to not touch file paths at all
> 
> How would you do this? Do you mean using `beet import -C ...`

Yes (but you can also set that permanently in the configuration: https://beets.readthedocs.io/en/stable/reference/config.html#importer-options). That way, you could move the files to the place you want them, and only then run the import (which will still add the track to the database, fetch metadata, and write that to the files).

**Reply by jameshibbard** [reply]
Got it. Thanks.
